File: VB/CustomBarTransparentButton/BarTransparentButtonItem.vb
' Developer Express Code Central Example:
' How to create a custom bar item with a transparent background
' 
' This sample demonstrates how to create a custom bar item with a transparent
' background. As an ancestor of a custom bar item the BarLargeButtonItem control
' has been chosen. This approach can be used to create items with a custom shape
' by adding images to the LargeGlyph, LargeGlyphHot and LargeGlyphPressed
' properties.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E2460

Imports System.ComponentModel
Imports System.Drawing
Imports System.Drawing.Design
Imports DevExpress.Utils
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Styles

Namespace CustomTransparentBarItem
    Public Class BarTransparentButtonItem
        Inherits BarLargeButtonItem

        Public Const BaseBarItemName As String = "BarLargeButtonItem"
        Public Const BarItemName As String = "BarTransparentButtonItem"
        Public Const BarItemCaption As String = "Transparent button"


        Private largeGlyphPressed_Renamed As Image

        Private largeImageIndexPressed_Renamed As Integer

        Shared Sub New()
            Register()
        End Sub
        Public Shared Sub Register()
            Register(BarAndDockingController.Default)
        End Sub
        Public Shared Sub Register(ByVal controller As BarAndDockingController)
            Register(controller.PaintStyles)
        End Sub
        Public Shared Sub Register(ByVal styles As BarManagerPaintStyleCollection)
            For i As Integer = 0 To styles.Count - 1
                Dim paintStyle As BarManagerPaintStyle = styles(i)
                Dim list As BarItemInfo = paintStyle.ItemInfoCollection(BaseBarItemName)
                If list IsNot Nothing AndAlso paintStyle.ItemInfoCollection(BarItemName) Is Nothing Then
                    paintStyle.ItemInfoCollection.Add(New BarItemInfo(BarItemName, BarItemCaption, -1, GetType(BarTransparentButtonItem), list.LinkType, GetType(BarTransparentButtonLinkViewInfo), list.LinkPainter, True, False))
                End If
            Next i
        End Sub

        Public Sub New()
            MyBase.New()
            largeGlyphPressed_Renamed = Nothing
            largeImageIndexPressed_Renamed = -1
        End Sub
        Public Sub New(ByVal manager As BarManager, ByVal caption As String)
            MyBase.New(manager, caption)
        End Sub
        Public Sub New(ByVal manager As BarManager, ByVal caption As String, ByVal imageIndex As Integer)
            MyBase.New(manager, caption, imageIndex)
        End Sub
        Public Sub New(ByVal manager As BarManager, ByVal caption As String, ByVal imageIndex As Integer, ByVal shortcut As BarShortcut)
            MyBase.New(manager, caption, imageIndex, shortcut)
        End Sub

        <Description("Gets or sets the large image displayed within associated links when they are pressed."), System.ComponentModel.DefaultValue(CType(Nothing, Object)), Category("Appearance")> _
        Public Overridable Property LargeGlyphPressed() As Image
            Get
                Return largeGlyphPressed_Renamed
            End Get
            Set(ByVal value As Image)
                If LargeGlyphHot Is value Then
                    Return
                End If
                largeGlyphPressed_Renamed = UpdateImage(value)
                OnItemChanged()
            End Set
        End Property

        <Description("Gets or sets the index of the large image displayed within associated links when they are pressed."), DefaultValue(-1), Category("Appearance"), Editor(GetType(DevExpress.Utils.Design.ImageIndexesEditor), GetType(UITypeEditor)), ImageList("LargeImages")> _
        Public Overridable Property LargeImageIndexPressed() As Integer
            Get
                Return largeImageIndexPressed_Renamed
            End Get
            Set(ByVal value As Integer)
                If LargeImageIndexHot = value Then
                    Return
                End If
                largeImageIndexPressed_Renamed = value
                OnItemChanged()
            End Set
        End Property
    End Class
End Namespace

File: CS/CustomBarTransparentButton/BarTransparentButtonItem.cs
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Design;
using DevExpress.Utils;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Styles;

namespace CustomTransparentBarItem
{
	public class BarTransparentButtonItem : BarLargeButtonItem
	{
		public const string BaseBarItemName = "BarLargeButtonItem";
		public const string BarItemName = "BarTransparentButtonItem";
		public const string BarItemCaption = "Transparent button";

		private Image largeGlyphPressed;
		private int largeImageIndexPressed;

		static BarTransparentButtonItem()
		{
			Register();
		}
		public static void Register()
		{
			Register(BarAndDockingController.Default);
		}
		public static void Register(BarAndDockingController controller)
		{
			Register(controller.PaintStyles);
		}
		public static void Register(BarManagerPaintStyleCollection styles)
		{
			foreach ( BarManagerPaintStyle paintStyle in styles )
			{
				BarItemInfo list = paintStyle.ItemInfoCollection[BaseBarItemName];
				if ( list != null && paintStyle.ItemInfoCollection[BarItemName] == null )
					paintStyle.ItemInfoCollection.Add(new BarItemInfo(BarItemName, BarItemCaption, -1, typeof(BarTransparentButtonItem), list.LinkType,
																		typeof(BarTransparentButtonLinkViewInfo), list.LinkPainter, true, false));
			}
		}

		public BarTransparentButtonItem()
			: base()
		{
			largeGlyphPressed = null;
			largeImageIndexPressed = -1;
		}
		public BarTransparentButtonItem(BarManager manager, string caption)
			: base(manager, caption)
		{
		}
		public BarTransparentButtonItem(BarManager manager, string caption, int imageIndex)
			: base(manager, caption, imageIndex)
		{
		}
		public BarTransparentButtonItem(BarManager manager, string caption, int imageIndex, BarShortcut shortcut)
			: base(manager, caption, imageIndex, shortcut)
		{
		}

		[Description("Gets or sets the large image displayed within associated links when they are pressed."), DefaultValue(null), Category("Appearance")]
		public virtual Image LargeGlyphPressed
		{
			get
			{
				return largeGlyphPressed;
			}
			set
			{
				if ( LargeGlyphHot == value ) return;
				largeGlyphPressed = UpdateImage(value);
				OnItemChanged();
			}
		}

		[Description("Gets or sets the index of the large image displayed within associated links when they are pressed."), DefaultValue(-1), Category("Appearance"), Editor(typeof(DevExpress.Utils.Design.ImageIndexesEditor), typeof(UITypeEditor)), ImageList("LargeImages")]
		public virtual int LargeImageIndexPressed
		{
			get
			{
				return largeImageIndexPressed;
			}
			set
			{
				if ( LargeImageIndexHot == value ) return;
				largeImageIndexPressed = value;
				OnItemChanged();
			}
		}
	}
}
